### DISCUSSION TITLE
A small discussion for Clip method

### DISCUSSION THREAD

**Discussion by citystrawman** [CGAL/cgal#8317] [General]
I am tweaking Visitor to understand how Clip method works. Here's my basic settings:
I have slope and slide as follows and I use `clip(slide, slope, params::visitor(), params::do_not_modify(false))` and after that both slide and slope got "cut", and slide surface mesh only remains the part that is inside slope. 
![image](https://github.com/CGAL/cgal/assets/9004748/b0df42ad-2f9e-4779-b1cc-ebce18a17a87)
![image](https://github.com/CGAL/cgal/assets/9004748/b2baf1f0-2edd-4366-b6dc-ba5b9a4455ae)

Then I use Visitor to identify the splitted halfedges and their target(vertex), and plot them by `add_property_map` . Here's my definition of visitor:
```
struct Visitor : public CGAL::Polygon_mesh_processing::Corefinement::Default_visitor<Mesh>
{
    std::vector<vertex_descriptor>& tmNewVertex;
    std::vector<vertex_descriptor>& splitterNewVertex;
    std::vector<halfedge_descriptor>& tmNewHalfEdge;
    std::vector<halfedge_descriptor>& splitterNewHalfEdge;
 
    Mesh& tm; 
    Mesh& splitter;

    Visitor(Mesh& tm, Mesh& splitter, std::vector<vertex_descriptor>& tmNewVertex, 
    std::vector<vertex_descriptor>& splitterNewVertex, std::vector<halfedge_descriptor>& tmNewHalfEdge, 
    std::vector<halfedge_descriptor>& splitterNewHalfEdge)
        : tm(tm), splitter(splitter), tmNewVertex(tmNewVertex), splitterNewVertex(splitterNewVertex), 
          tmNewHalfEdge(tmNewHalfEdge), splitterNewHalfEdge(splitterNewHalfEdge) {}

    void edge_split(halfedge_descriptor  hnew , Mesh&  tm ) 
    {
        std::cout << "edge_split called: hnew: " << hnew  << ", hnew's incident vertex: " << tm.target(hnew) << std::endl;
        if (&this->tm == std::addressof(tm)) {
            tmNewHalfEdge.push_back(hnew);
        }
        else if (&this->splitter == std::addressof(tm)) {
            splitterNewHalfEdge.push_back(hnew);
        }
    }
}
```
From the visitor I am trying to use `edge_split` to collect the new halfedges that are created when doing clipping.
After that, I am going to colorize the halfedges **and their corresponding vertex**:

```
auto vcm1 = slide.add_property_map<Mesh::Vertex_index, CGAL::IO::Color>("v:color").first;
auto vcm2 = slope.add_property_map<Mesh::Vertex_index, CGAL::IO::Color>("v:color").first;
auto ecm1 = slide.add_property_map<Mesh::edge_index, CGAL::IO::Color>("e:color").first;
auto ecm2 = slope.add_property_map<Mesh::edge_index, CGAL::IO::Color>("e:color").first;

for each (auto halfEdge in tmNewHalfEdge)
{
    auto edge = slide.edge(halfEdge);
    put(ecm1, edge, CGAL::IO::red());

    auto vertex = target(halfEdge, slide);
    put(vcm1, vertex, CGAL::IO::red());
}

for each (auto halfEdge in splitterNewHalfEdge)
{
    auto edge = slope.edge(halfEdge);
    put(ecm2, edge, CGAL::IO::red());

    auto vertex = target(halfEdge, slope);
    put(vcm2, vertex, CGAL::IO::red());
}
```

then I visualize the clipped slide and slope: 

![image](https://github.com/CGAL/cgal/assets/9004748/b7c66ad8-8ccf-4d18-9ded-af38b8791ea6)

![image](https://github.com/CGAL/cgal/assets/9004748/589e131d-2466-454a-ac38-77a02479bc29)

what confused me, is that for slide, there's some vertex which does not have its corresponding halfedge shown. I guess, probably, these halfedges actually exists in the removed part But from the code, the vertex is obtained by the halfedge and slide, does that mean that slide still maintains the halfedges that are removed ?

**Comment by sloriot** [comment]
Probably missing [this one](https://doc.cgal.org/latest/Polygon_mesh_processing/classPMPCorefinementVisitor.html#ab819d40a2bda062d883ac2a6fb62e8f8).

I don't understand why the solution I provided with corefinement in one of your tickets is not doing what you want. Maybe post a 2D picture of the input and what output you'd like to have.

**Reply by citystrawman** [reply]
> Probably missing [this one](https://doc.cgal.org/latest/Polygon_mesh_processing/classPMPCorefinementVisitor.html#ab819d40a2bda062d883ac2a6fb62e8f8).
> 
> I don't understand why the solution I provided with corefinement in one of your tickets is not doing what you want. Maybe post a 2D picture of the input and what output you'd like to have.

Hello, please refer to this post: https://github.com/MeshInspector/MeshLib/issues/2899 
This post shows what I want to do.  I appologize for unable to express clearly.
